### PR TITLE
SP-6203-store-expose-step-exeucution-error

### PR DESCRIPTION
**Pull request by talbe**

### CODE REVIEW

File: steps/redis/get/main.go [hunk 1]
```diff
@@ -21,43 +24,56 @@ type RedisGet struct {
 	redisClient *redis.Client
 }
 
-func (l *RedisGet) Init() error {
+func (l *RedisGet) Init() *step.Error {
 	err := envconf.Parse(&l.args)
 	if err != nil {
-		return err
+		return &step.Error{Code: step.InvalidArgument, Message: "invalid env vars provided", Verbose: fmt.Errorf("error: %w", err).Error()}
```
Comment (kostyay) [inline steps/redis/get/main.go:30]:
```suggestion
		return &step.Error{Code: step.InvalidArgument, Message: "Environment variables are missing or invalid", Verbose: fmt.Errorf("error: %w", err).Error()}
```

File: steps/redis/get/main.go [hunk 2]
```diff
@@ -21,43 +24,56 @@ type RedisGet struct {
...
 	}
 
 	redisClient, err := base.InitRedisClient(l.args.Args)
 	if err != nil {
-		return err
+		return &step.Error{Code: step.Unavailable, Message: "failed connect redis client", Verbose: fmt.Errorf("error: %w", err).Error()}
```
Comment (kostyay) [inline steps/redis/get/main.go:35]:
```suggestion
		return &step.Error{Code: step.Unavailable, Message: "Failed connecting to redis server", Verbose: fmt.Errorf("error: %w", err).Error()}
```

File: steps/redis/get/main.go [hunk 3]
```diff
@@ -21,43 +24,56 @@ type RedisGet struct {
...
 }
 
-func (l *RedisGet) Run() (exitCode int, output []byte, err error) {
+func (l *RedisGet) Run() (outputJson step.Output, stepError *step.Error) {
 	gc := gabs.New()
 
 	cmd := l.redisClient.Get(context.Background(), l.args.Key)
 	if cmd == nil {
-		return step.ExitCodeFailure, nil, fmt.Errorf("invalid result returned from get operation")
+		return nil, &step.Error{Code: step.InvalidArgument, Message: "invalid result returned from get operation"}
```
Comment (kostyay) [inline steps/redis/get/main.go:48]: im not sure InvalidArgument is the right status for this error
maybe Internal?

File: steps/redis/get/main.go [hunk 4]
```diff
@@ -21,43 +24,56 @@ type RedisGet struct {
...
 	}
 
 	val, err := cmd.Result()
 	if err != nil {
 		if err == redis.Nil {
-			return step.ExitCodeFailure, nil, fmt.Errorf("key not found")
+			return nil, &step.Error{Code: step.NotFound, Message: "key not found"}
 		}
-		return step.ExitCodeFailure, nil, fmt.Errorf("failed getting key value: validate key is valid: %w", err)
+		return nil, &step.Error{Code: step.InvalidArgument, Message: "failed getting key value", Verbose: fmt.Sprintf("validate key is valid: %w", err)}
```
Comment (kostyay) [inline steps/redis/get/main.go:56]:
```suggestion
		return nil, &step.Error{Code: step.InvalidArgument, Message: "failed getting key value", Verbose: fmt.Sprintf("validate key is valid: %s", err.Error())}
```
Cant use %w outside of fmt.Errorf

File: steps/redis/get/main.go [hunk 5]
```diff
@@ -21,43 +24,56 @@ type RedisGet struct {
...
 
-	return step.ExitCodeOK, gc.Bytes(), nil
+	var output map[string]interface{}
+	json.Unmarshal(gc.Bytes(), &output)
+	if err != nil {
+		return map[string]interface{}{}, &step.Error{
+			Code:    step.InvalidArgument,
+			Message: "bad output",
+			Verbose: fmt.Errorf("output should be json. error: %w", err).Error()}
+	}
```
Comment (kostyay) [inline steps/redis/get/main.go:68]: what is the purpose of this? i dont think that u need it

**Review by kostyay — CHANGES_REQUESTED** [review]
see comments

File: steps/redis/get/main.go [hunk 6]
```diff
@@ -21,43 +24,56 @@ type RedisGet struct {
...
+		return &step.Error{Code: step.Unavailable, Message: "failed connect redis client", Verbose: fmt.Errorf("error: %w", err).Error()}
 	}
 
 	l.redisClient = redisClient
 
 	return nil
 }
 
-func (l *RedisGet) Run() (exitCode int, output []byte, err error) {
+func (l *RedisGet) Run() (outputJson step.Output, stepError *step.Error) {
```
Comment (kostyay) [inline steps/redis/get/main.go:43]:
```suggestion
func (l *RedisGet) Run() (step.Output, *step.Error) {
```

File: steps/redis/get/Dockerfile [hunk 7]
```diff
@@ -28,4 +28,4 @@ RUN chmod +x ${BINARY_NAME}
 COPY --from=base /usr/local/bin/ /usr/local/bin/
 
 ENV STEP_NAME=${BINARY_NAME}
-ENTRYPOINT ["/bin/sh", "-c", "set -eo pipefail; /${STEP_NAME} 2>&1 | sp-base-step format"]
\ No newline at end of file
+ENTRYPOINT ["/bin/sh", "-c", "/${STEP_NAME}"]
```
Comment (hagaishapira) [inline steps/redis/get/Dockerfile:31]: Should we do this change to all dockerfiles who use the new sdk?

**Review by hagaishapira — COMMENTED** [review]
Please also update the cookiecutter template located at scripts/step-template

File: steps/redis/get/Dockerfile (hunk 7 above)
Comment (hagaishapira) [reply steps/redis/get/Dockerfile:31]: @kostyay @talbe ?

Comment (talbe) [reply steps/redis/get/Dockerfile:31]: @hagaishapira Yeah, sp-base-step is redundant

**Review by hagaishapira — APPROVED** [review]
LGTM

**Pull request merged**